### PR TITLE
PCA: Implement computeCorrelationEigenvalues on GPU

### PR DESCRIPTION
**Pull request by mchernov-intel**

### CODE REVIEW

**Comment by mchernov-intel** [comment]
make test

**Comment by mchernov-intel** [comment]
make test

File: algorithms/kernel/pca/oneapi/cl_kernels/pca_cl_kernels.cl [hunk 1]
```diff
@@ -38,6 +38,20 @@ DECLARE_SOURCE(
...
     }
 
+    __kernel void sortEigenvectorsDescending(__global algorithmFPType * fullEigenvectors, __global algorithmFPType * fullEigenvalues) {
+        const int nFeatures = get_global_size(1);
+        const int i = get_global_id(0);
+        const int j = get_global_id(1);
+
+        algorithmFPType tmp = fullEigenvalues[i];
+        fullEigenvalues[i]                 = fullEigenvalues[nFeatures - 1 - i];
+        fullEigenvalues[nFeatures - 1 - i] = tmp;
```
Comment (mchernov-intel) [inline algorithms/kernel/pca/oneapi/cl_kernels/pca_cl_kernels.cl]: It doesn't work if nFeatures is even. I gonna fix it.

File: include/oneapi/internal/math/mkl_lapack.h [hunk 2]
```diff
@@ -103,12 +103,49 @@ struct MKLPotrs
...
+        cl::sycl::buffer<algorithmFPType, 1> work_sycl_buff = work.toSycl();
+        cl::sycl::buffer<int64_t, 1> iwork_sycl_buff        = iwork.toSycl();
+
+        cl::sycl::buffer<int64_t, 1> info(cl::sycl::range<1>(1));
+        fpk::lapack::syevd(_queue, jobmkl, uplomkl, n, a_sycl_buff, lda, w_sycl_buff, work_sycl_buff, lwork, iwork_sycl_buff, liwork, info);
+
+        _queue.wait();
+        /** TODO: Check info buffer for containing errors. Now it is not supported.:
+         *  https://software.intel.com/en-us/oneapi-mkl-dpcpp-developer-reference-potrs
+        */
```
Comment (michael-smirnov) [inline include/oneapi/internal/math/mkl_lapack.h]: fix this todo

File: include/oneapi/internal/math/mkl_lapack.h [hunk 3]
```diff
@@ -103,12 +103,49 @@ struct MKLPotrs
...
+    MKLXsyevd(cl::sycl::queue & queue) : _queue(queue) {}
+
+    services::Status operator()(const math::Job jobz, const math::UpLo uplo, const int64_t n, services::Buffer<algorithmFPType> & a,
+                                const int64_t lda, services::Buffer<algorithmFPType> & w, services::Buffer<algorithmFPType> & work,
+                                const int64_t lwork, services::Buffer<int64_t> & iwork, const int64_t liwork)
+    {
+        services::Status status;
+
+        const fpk::uplo uplomkl                             = uplo == math::UpLo::Upper ? fpk::uplo::upper : fpk::uplo::lower;
+        const fpk::job jobmkl                               = fpk::job(jobz);
```
Comment (michael-smirnov) [inline include/oneapi/internal/math/mkl_lapack.h]: Why you are converting these enums in different ways?

File: service/kernel/oneapi/lapack_gpu.cpp [hunk 4]
```diff
@@ -79,12 +80,64 @@ services::Status ReferencePotrs<algorithmFPType>::operator()(const math::UpLo up
...
+    case math::Job::overwritevec: job = 'O'; break;
+
+    default: status.add(services::UnknownError); break;
+    }
+    DAAL_CHECK_STATUS_VAR(status);
+
+    DAAL_INT nInt      = static_cast<DAAL_INT>(n);
+    DAAL_INT ldaInt    = static_cast<DAAL_INT>(lda);
+    DAAL_INT lworkInt  = static_cast<DAAL_INT>(lwork);
+    DAAL_INT liworkInt = static_cast<DAAL_INT>(liwork);
```
Comment (michael-smirnov) [inline service/kernel/oneapi/lapack_gpu.cpp]: check SDL about these conversions

File: service/kernel/oneapi/lapack_gpu.cpp [hunk 5]
```diff
@@ -79,12 +80,64 @@ services::Status ReferencePotrs<algorithmFPType>::operator()(const math::UpLo up
...
+    services::internal::TArray<DAAL_INT, CpuType::sse2> iWorkTmp(iwork.size());
+
+    LapackAutoDispatch<algorithmFPType>::xsyevd(&job, &up, &nInt, aPtr.get(), &ldaInt, wPtr.get(), workPtr.get(), &lworkInt, iWorkTmp.get(),
+                                                &liworkInt, &info);
+    DAAL_CHECK(info == 0, services::ErrorID::ErrorNormEqSystemSolutionFailed);
+
+    int64_t* iworkPtr = iworkSharedPtr.get();
+    for (size_t i = 0; i < iwork.size(); ++i) {
+        iworkPtr[i] = static_cast<int64_t>(iWorkTmp[i]);
+    }
```
Comment (michael-smirnov) [inline service/kernel/oneapi/lapack_gpu.cpp]: can we avoid such conversions? Does Lapack have different APIs for GPU and CPU?

File: algorithms/kernel/pca/oneapi/pca_dense_correlation_batch_kernel_ucapi_impl.i [hunk 6]
```diff
@@ -233,6 +234,106 @@ services::Status PCACorrelationKernelUCAPI<algorithmFPType>::calculateVariances(
...
+    const math::Job jobz  = math::Job::vec;
+    const math::UpLo uplo = math::UpLo::Upper;
+
+    const size_t lwork  = 2 * nFeatures * nFeatures + 6 * nFeatures + 1;
+    const size_t liwork = 5 * nFeatures + 3;
+
+    auto work = context.allocate(TypeIds::id<algorithmFPType>(), lwork, &status);
+    DAAL_CHECK_STATUS_VAR(status);
+    auto iwork = context.allocate(TypeIds::id<int64_t>(), liwork, &status);
+    DAAL_CHECK_STATUS_VAR(status);
```
Comment (michael-smirnov) [inline algorithms/kernel/pca/oneapi/pca_dense_correlation_batch_kernel_ucapi_impl.i]: What the purpose of this auxiliary staff?

File: service/kernel/oneapi/lapack_gpu.cpp (hunk 5 above)
Comment (mchernov-intel) [reply service/kernel/oneapi/lapack_gpu.cpp]: In the end CPU requires `DAAL_INT *`, but GPU needs `int64_t`

Comment (michael-smirnov) [reply service/kernel/oneapi/lapack_gpu.cpp]: Which type is wider?

File: algorithms/kernel/pca/oneapi/pca_dense_correlation_batch_kernel_ucapi_impl.i (hunk 6 above)
Comment (mchernov-intel) [reply algorithms/kernel/pca/oneapi/pca_dense_correlation_batch_kernel_ucapi_impl.i]: align to CPU

**Comment by mchernov-intel** [comment]
make test

**Comment by mchernov-intel** [comment]
restart

**Comment by mchernov-intel** [comment]
make test

File: algorithms/kernel/pca/oneapi/cl_kernels/pca_cl_kernels.cl [hunk 7]
```diff
@@ -38,6 +38,19 @@ DECLARE_SOURCE(
         variances[tid] = covariance[tid * nFeatures + tid];
     }
 
+    __kernel void sortEigenvectorsDescending(__global algorithmFPType * fullEigenvectors, __global algorithmFPType * fullEigenvalues,
```
Comment (michael-smirnov) [inline algorithms/kernel/pca/oneapi/cl_kernels/pca_cl_kernels.cl]:
```suggestion
    __kernel void sortEigenvectorsDescending(__global const algorithmFPType * fullEigenvectors, __global const algorithmFPType * fullEigenvalues,
```

File: algorithms/kernel/pca/oneapi/pca_dense_correlation_batch_kernel_ucapi_impl.i (hunk 6 above)
Comment (michael-smirnov) [reply algorithms/kernel/pca/oneapi/pca_dense_correlation_batch_kernel_ucapi_impl.i]: Can you explain to me what `work`, `lwork`, `iwork` and `liwork` are used for?

File: service/kernel/service_data_utils.h [hunk 8]
```diff
@@ -25,6 +25,9 @@
 #define __SERVICE_DATA_UTILS_H__
 
 #include "service/kernel/service_defines.h"
+#include <climits>
+#include <cfloat>
+#include <limits>
```
Comment (michael-smirnov) [inline service/kernel/service_data_utils.h]: forgotten includes?

Comment (mchernov-intel) [reply service/kernel/service_data_utils.h]: This includes are important for correct building. Constants used in this header are not redefined.

**Comment by mchernov-intel** [comment]
make test

**Comment by mchernov-intel** [comment]
make test

**Comment by mchernov-intel** [comment]
make test

**Comment by mchernov-intel** [comment]
make test

**Comment by mchernov-intel** [comment]
make test

**Pull request closed without merging**